File: app/schemas/tutor.py
"""Pydantic schemas for the AI Tutor API."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


# ── Request schemas ──────────────────────────────────────────────────────────

class AnswerSubmission(BaseModel):
    """Request body for POST /answer."""
    student_answer: str


# ── Response schemas ─────────────────────────────────────────────────────────

class QuestionOut(BaseModel):
    """Response for GET /question."""
    id: str
    subject: str
    topic: str
    difficulty: str
    question_text: str
    knowledge_tags: list[str]
    learning_objectives: list[str]


class TutorResponseOut(BaseModel):
    """Pedagogical feedback from the LLM tutor."""
    explanation: str
    hint: str
    encouragement: str


class DiagnosisResultOut(BaseModel):
    """Error diagnosis result."""
    error_types: list[str]
    diagnosis_labels: list[str]
    confidence: float


class RemediationPlanOut(BaseModel):
    """Remediation plan with recommended topics."""
    recommended_topics: list[str]
    retrieval_tags: list[str]


class TutoringExplanationOut(BaseModel):
    """Plain-language explanation for a wrong answer."""
    what_is_wrong: str
    why_it_is_wrong: str
    how_to_fix: str
    similar_examples: list[str]
    retrieved_context: str
    generation_source: Optional[str] = None
    metadata: dict[str, Any] = {}


class KnowledgeSnippetOut(BaseModel):
    """A retrieved knowledge document with metadata."""
    id: str
    title: str
    topic: str
    tags: list[str]
    diagnosis_labels: list[str]
    score: float
    metadata: dict[str, Any] = {}


class AnswerResult(BaseModel):
    """Response for POST /answer."""
    is_correct: bool
    correct_answer: str
    student_answer: str
    topic: str
    tutor_response: Optional[TutorResponseOut] = None
    diagnosis: Optional[DiagnosisResultOut] = None
    remediation: Optional[RemediationPlanOut] = None
    explanation: Optional[TutoringExplanationOut] = None
    retrieved_snippets: list[KnowledgeSnippetOut] = []


class HintResponse(BaseModel):
    """Response for POST /hint."""
    hint: str
    level: int
    remaining: int


class MasteryStateOut(BaseModel):
    """Per-topic mastery state."""
    topic_id: str
    alpha: float
    beta: float
    total_attempts: int
    correct_attempts: int
    last_seen: Optional[datetime] = None
    mastery: float
    variance: float


class ProgressOut(BaseModel):
    """Full session progress snapshot."""
    mastery_states: list[MasteryStateOut]
    total_questions: int
    correct_count: int
    accuracy: float
    correct_streak: int
    wrong_streak: int

File: frontend/src/shared/api/tutor.ts
import { api } from '@/lib/api';

// 1:1 mirror of backend Pydantic schemas in `app/schemas/tutor.py` and the
// response models defined in `app/api/v1/tutor_router.py`.
// Field names kept in sync with the backend.

export interface Question {
  id: string;
  subject: string;
  topic: string;
  difficulty: string;
  question_text: string;
  knowledge_tags: string[];
  learning_objectives: string[];
}

// Backend has no `options` field on the question. The UI synthesizes an
// A/B/C/D option set so the existing multiple-choice QuestionCard still
// renders. The chosen option letter is sent as `student_answer` to the
// backend, which compares it to the actual text answer returned in
// `AnswerResult.correct_answer`.
export interface SubmitResult {
  is_correct: boolean;
  correct_answer: string;
  student_answer: string;
  topic: string;
  explanation: string;
  hint?: string | null;
  encouragement?: string | null;
}

export interface HintResponse {
  hint: string;
  level: number;
  remaining: number;
}

export interface MasteryState {
  topic_id: string;
  alpha: number;
  beta: number;
  total_attempts: number;
  correct_attempts: number;
  last_seen: string | null;
  mastery: number;
  variance: number;
}

export interface Progress {
  mastery_states: MasteryState[];
  total_questions: number;
  correct_count: number;
  accuracy: number;
  correct_streak: number;
  wrong_streak: number;
}

export interface SubjectInfo {
  subject: string;
  available_subjects: string[];
}

export const tutorApi = {
  getNextQuestion: () => api.get<Question>('/tutor/question'),

  // The backend's `AnswerSubmission` is just `{ student_answer: string }`.
  // We accept an `optionId` from the UI and forward it as `student_answer`.
  submitAnswer: (data: { optionId: string }) =>
    api.post<SubmitResult>('/tutor/answer', { student_answer: data.optionId }),

  requestHint: () => api.post<HintResponse>('/tutor/hint', {}),

  getMastery: () => api.get<MasteryState[]>('/tutor/mastery'),

  getProgress: () => api.get<Progress>('/tutor/progress'),

  getSubject: () => api.get<SubjectInfo>('/tutor/subject'),

  switchSubject: (subject: string) =>
    api.post<SubjectInfo>('/tutor/subject', { subject }),

  reset: () => api.post<{ status: string }>('/tutor/reset', {}),
};
